src/vmm/src/signal_handler.rs

```diff
@@ -258,27 +236,26 @@ mod tests {
         });
         assert!(child.join().is_ok());
 
-        // Sanity check.
-        assert!(cpu_count() > 0);
-        // Kcov somehow messes with our handler getting the SIGSYS signal when a bad syscall
-        // is caught, so the following assertion no longer holds. Ideally, we'd have a surefire
-        // way of either preventing this behaviour, or detecting for certain whether this test is
-        // run by kcov or not. The best we could do so far is to look at the perceived number of
-        // available CPUs. Kcov seems to make a single CPU available to the process running the
-        // tests, so we use this as an heuristic to decide if we check the assertion.
-        if cpu_count() > 1 {
-            // The signal handler should let the program continue during unit tests.
-            assert!(METRICS.seccomp.num_faults.fetch() >= 1);
-        }
+        // SIGSYS, which is raised whenever a bad syscall is caught will be intercepted by kcov on x86_64
+        // and thus not reach Firecracker:
+        // https://github.com/SimonKagstrom/kcov/blob/a8b60c43fb33f56553a2bb20633e3b59a08abae1/src/engines/ptrace.cc#L187
+        // So, we are not checking for the `num_faults` metrics which gets incremented on each bad syscall
+        // if we run with kcov and we are on x86_64.
+        assert!(
+            METRICS.seccomp.num_faults.fetch()
+                >= (!run_with_kcov || !cfg!(target_arch = "x86_64")) as usize
```

Comment:
Do we actually need to check for the `target_arch` here? There wasn't there in the original code.

Reply:
Yes, so I wanted to be really strict about this in order to reflect on the reason behind why this happens. The num_faults is 0 only on x86_64 since kcov will intercept the SIGSYS while on ARM, SIGILL is the one intercepted which motivates the next change.
The explanation is here: https://github.com/SimonKagstrom/kcov/blob/a8b60c43fb33f56553a2bb20633e3b59a08abae1/src/engines/ptrace.cc#L187